### DISCUSSION TITLE
How to avoid duplication between experiments

### DISCUSSION THREAD

**Discussion by DirkKuhn** [mit-ll-responsible-ai/hydra-zen#504] [Q&A]
Hey,

hydra_zen has helped me a lot. Right now I have many experiments and I was wondering how I could reduce duplication between them.
Consider the following toy example.
```python
from hydra_zen import store, builds, MISSING, make_config, zen

# Toy application
class Item:
    def __init__(self, data: int):
        self.data = data

    def __repr__(self) -> str:
        return f'Item(data={self.data})'

class Test:
    def __init__(self, data: int):
        self.data = data

    def __repr__(self) -> str:
        return f'Item(data={self.data})'

def app(item: Item, test: Test, number: int, msg: str) -> None:
    print(f'{item=}')
    print(f'{test=}')
    print(f'{number=}')
    print(f'{msg=}')


# Configuration
item_store = store(group='item')
item_store(Item, data=1)

test_store = store(group='test')
test_store(Test, data=1, name='test1')
test_store(Test, data=2, name='test2')

# I do not want to extract duplicated values here
Config = builds(
    app,
    hydra_defaults=[
        '_self_',
        {'item': MISSING},
        {'test': MISSING},
        {'experiment': 'experiment1'}
    ],
    populate_full_signature=True
)
store(Config, name='config')

# 2 experiments with considerable duplication
experiment_store = store(group='experiment', package='_global_')
experiment_store(
    make_config(
        hydra_defaults=[
            '_self_',
            {'override /item': 'Item'},
            {'override /test': 'test1'},
        ],
        number=10,
        msg='m1',
        bases=(Config,)
    ), name='experiment1'
)
experiment_store(
    make_config(
        hydra_defaults=[
            '_self_',
            {'override /item': 'Item'},
            {'override /test': 'test2'},
        ],
        number=10,
        msg='m2',
        bases=(Config,)
    ), name='experiment2'
)

if __name__ == '__main__':
    store.add_to_hydra_store()
    zen(app).hydra_main(
        config_name='config',
        config_path=None,
        version_base='1.3'
    )
```
Both experiment1 and experiment2 use the config group `'Item'` und set `number` to 10. How can this be shared?
I also do not want to put these values into `Config` as there might be many more experiments which use a different `Item` and `number`.

Thank you so much!

**Comment by rsokl** [comment]
What do you think of this?

```python
experiment_store = store(group="experiment", package="_global_")
Shared = make_config(
    hydra_defaults=[
        "_self_",
        {"override /item": "Item"},
        {"override /test": "test1"},
    ],
    number=10,
    msg=MISSING,
    bases=(Config,),
)


experiment_store(make_config(msg="m1", bases=(Shared,)), name="experiment1")
experiment_store(make_config(msg="m2", bases=(Shared,)), name="experiment2")
```

Using another level of inheritance seems like it could be a good approach for sharing parameters across experiments

**Reply by DirkKuhn** [reply]
I like your suggestion. Unfortunately, I am running into some problems.
If in 'experiment2' I want to change the the `hydra_defaults` like that
```python
experiment_store(
    make_config(
        hydra_defaults=[
            '_self_',
            {'override /test': 'test2'},
        ],
        msg='m2',
        bases=(Shared,)
    ), name='experiment2'
)
```
I get the error
```
You must specify 'item', e.g, item=<OPTION>
Available options:
	Item
```
It seems the `hydra_defaults` of `Shared` are simply replaced.
Do you have an idea how to mitigate this?

**Reply by DirkKuhn** [reply]
For now I am using
```python
shared_defaults = [{'override /item': 'Item'}]
Shared = make_config(
    number=10,
    msg=MISSING,
    bases=(Config,),
)

experiment_store = store(group='experiment', package='_global_')
experiment_store(
    make_config(
        hydra_defaults=[
            '_self_',
            *shared_defaults,  # Shared config groups
            {'override /test': 'test1'},
        ],
        msg='m1',
        bases=(Shared,)  # Shared values
    ), name='experiment1'
)
```
it is not pretty but it works.
I really like the `*` expression in python :)
I also tried putting the shared values in a separate config group in the global package but to no avail.

**Reply by jgbos** [reply]
I don't see any issue with your solution, but could you not just set a default for `item` in the main `Config` and have experiments override when needed?

Example updates to your code if you want the experiment to update "item":

```python
item_store(Item, data=2, name="OverrideItem")

Config = builds(
    app,
    hydra_defaults=[
        '_self_',
        {"item": "Item"},
        {'test': MISSING},
        {'experiment': 'experiment1'}
    ],
    populate_full_signature=True
)

experiment_store(
    make_config(
        hydra_defaults=[
            '_self_',
            {"override /item": "OverrideItem"},
            {'override /test': 'test2'},
        ],
        msg='m2',
        bases=(Shared,)
    ), name='experiment2'
)
```

**Reply by jgbos** [reply]
Also, not sure if this is helpful to your question, but you can set defaults for "sub-configs".  Take a look at the hydra documentation here:

https://hydra.cc/docs/advanced/defaults_list/
https://hydra.cc/docs/patterns/select_multiple_configs_from_config_group/

**Reply by DirkKuhn** [reply]
@jgbos Thank you for your answer. Admittedly, this toy example is too small to visualize the problem of duplication.

I would prefer not to set the default of `item` in `Config` as this might be a bit missleading if I have completely different experiments.

Furthermore, I would like to avoid specifying shared defaults in each experiment separately. Here, it is no problem to set `{"override /item": "OverrideItem"}` but what if I also have `experiment3` - `experiment10` all with 10 shared defaults?

I am not sure what you mean by "sub-configs". Can you give an example?
